include/veriblock/bfi/bitcoin/transaction.hpp

```diff
@@ -0,0 +1,212 @@
+};
+
+/** An output of a transaction.  It contains the public key that the next input
+ * must be able to sign with to claim it.
+ */
+struct TxOut {
+  Amount nValue;
+  Script scriptPubKey;
+
+  ADD_SERIALIZE_METHODS;
+
+  template <typename Stream, typename Operation>
+  inline void SerializationOp(Stream& s, Operation ser_action) {
+    READWRITE(nValue);
+    READWRITE(scriptPubKey);
+  }
+
+  friend bool operator==(const TxOut& a, const TxOut& b) {
+    return a.nValue == b.nValue && a.scriptPubKey == b.scriptPubKey;
+  }
+
+  friend bool operator!=(const TxOut& a, const TxOut& b) { return !(a == b); }
+};
+
+/** The basic transaction that is broadcasted on the network and contained in
+ * blocks.  A transaction can contain multiple inputs and outputs.
+ */
+struct Transaction {
+  // The local variables are made const to prevent unintended modification
+  // without updating the cached hash value. However, CTransaction is not
+  // actually immutable; deserialization and assignment are implemented,
+  // and bypass the constness. This is safe, as they update the entire
+  // structure, including the hash.
+  std::vector<TxIn> vin;
+  std::vector<TxOut> vout;
+  int32_t nVersion;
+  uint32_t nLockTime;
+
+  template <typename Stream>
+  inline void Serialize(Stream& s) const {
+    SerializeTransaction(*this, s);
+  }
+
+  template <typename Stream>
+  inline void Unserialize(Stream& s) {
+    UnserializeTransaction(*this, s);
+  }
+
+  friend bool operator==(const Transaction& a, const Transaction& b) {
+    return a.vin == b.vin && a.vout == b.vout && a.nVersion == b.nVersion &&
+           a.nLockTime == b.nLockTime;
+  }
+
+  friend bool operator!=(const Transaction& a, const Transaction& b) {
+    return !(a == b);
+  }
+
+  bool HasWitness() const {
+    for (size_t i = 0; i < vin.size(); i++) {
+      if (!vin[i].scriptWitness.empty()) {
+        return true;
+      }
+    }
+    return false;
+  }
+};
+
+template <typename Stream>
+inline void UnserializeTransaction(Transaction& tx, Stream& s) {
+  // TODO: fix
+  //    const bool fAllowWitness = !(s.GetVersion() &
+  //    SERIALIZE_TRANSACTION_NO_WITNESS);
+  const bool fAllowWitness = false;
+
+  UnserializeBtc(s, tx.nVersion);
+  unsigned char flags = 0;
+  tx.vin.clear();
+  tx.vout.clear();
+  /* Try to read the vin. In case the dummy is there, this will be read as an
+   * empty vector. */
+  UnserializeBtc(s, tx.vin);
+  if (tx.vin.size() == 0 && fAllowWitness) {
+    /* We read a dummy or an empty vin. */
+    UnserializeBtc(s, flags);
+    if (flags != 0) {
+      UnserializeBtc(s, tx.vin);
+      UnserializeBtc(s, tx.vout);
+    }
+  } else {
+    /* We read a non-empty vin. Assume a normal vout follows. */
+    UnserializeBtc(s, tx.vout);
+  }
+  if ((flags & 1) && fAllowWitness) {
+    /* The witness flag is present, and we support witnesses. */
+    flags ^= 1;
+    for (size_t i = 0; i < tx.vin.size(); i++) {
+      UnserializeBtc(s, tx.vin[i].scriptWitness);
```

Comment:
Is it necessary to call this function UnserializeBtc? Maybe just Unserialize?

Reply:
Hmm, good point will try